apple/internal/apple_framework_import.bzl

```diff
@@ -58,6 +58,26 @@ def _is_swiftmodule(path):
     """Predicate to identify Swift modules/interfaces."""
     return path.endswith(".swiftmodule") or path.endswith(".swiftinterface")
 
+def _swiftmodules_for_cpu(swiftmodule_files, cpu):
+    """Select cpu specific swiftmodules"""
+    # Assume that one Swift module or interface is valid.
+    if len(swiftmodule_files) == 1:
+        return swiftmodule_files
+
+    # Filter to specific architectures, for example "x86_64.swiftmodule".
+    cpu_prefix = cpu + "."
+    matches = [
+        f
+        for f in swiftmodule_files
+        if f.basename.startswith(cpu_prefix)
```

Comment:
is this format a requirement or a standard? could we find swift modules that do not follow this naming convention?

Reply:
Requirement, I'm not sure. Standard? Seems so. The swift repo testing declares a variable for this purpose: `target-swiftmodule-name`.

https://github.com/apple/swift/blob/cd84bc17ba5bce7cef6e560c4930644f44d46625/test/lit.cfg#L1612

And from [docs/Testing.md](https://github.com/apple/swift/blob/cd84bc17ba5bce7cef6e560c4930644f44d46625/docs/Testing.md):

> `%target-swiftmodule-name` and `%target-swiftdoc-name`: the basename of swiftmodule and swiftdoc files for a framework compiled for the target (for example, `arm64.swiftmodule` and `arm64.swiftdoc`).

Reply:
It looks like the future is getting more complicated than this.

The code that handles module loading is in https://github.com/apple/swift/blob/master/lib/Serialization/SerializedModuleLoader.cpp.

For frameworks, it looks like `$MODULE_NAME.swiftmodule/$ARCH.swiftmodule` is somewhat [legacy behavior](https://github.com/apple/swift/blob/master/lib/Serialization/SerializedModuleLoader.cpp#L43-L45).

The [modern behavior](https://github.com/apple/swift/blob/master/lib/Serialization/SerializedModuleLoader.cpp#L40) (which looks like it may have been introduced in Swift 5.1) does normalization with the full triple, not just the architecture, then uses the above as a fallback. That normalization is implemented [here](https://github.com/apple/swift/blob/master/lib/Basic/Platform.cpp#L310), and generates names like `x86_64-apple-ios-simulator.swiftmodule`.

There's also an [edge case](https://github.com/apple/swift/blob/master/lib/Serialization/SerializedModuleLoader.cpp#L47-L53) where the compiler will look for `arm.swiftmodule` instead of a more specific `arm64`, `armv7`, etc. variant.

So, depending on who builds the framework in question, if it emits modules with the newer names, it looks like this code may fail to find them.

Reply:
Great info thanks Tony.

_For future reference_, (since it's good documentation) here are the same links but pointing to the commits at this time:

1. [lib/Serialization/SerializedModuleLoader.cpp](https://github.com/apple/swift/blob/ca1348e6ae2ae872b92b8fc1cb7e3febb7363f4e/lib/Serialization/SerializedModuleLoader.cpp)
2. [legacy behavior](https://github.com/apple/swift/blob/ca1348e6ae2ae872b92b8fc1cb7e3febb7363f4e/lib/Serialization/SerializedModuleLoader.cpp#L43-L45)
3. [modern behavior](https://github.com/apple/swift/blob/ca1348e6ae2ae872b92b8fc1cb7e3febb7363f4e/lib/Serialization/SerializedModuleLoader.cpp#L40)
4. [normalization](https://github.com/apple/swift/blob/ca1348e6ae2ae872b92b8fc1cb7e3febb7363f4e/lib/Basic/Platform.cpp#L310)
5. [edge case](https://github.com/apple/swift/blob/ca1348e6ae2ae872b92b8fc1cb7e3febb7363f4e/lib/Serialization/SerializedModuleLoader.cpp#L47-L53)

Reply:
It's too bad the `ld` can't be given a `.swiftmodule` bundle, and use this logic themselves. I think I'll file a FB for this.

Reply:
In our case, we have our own rule that generates `.swiftmodule` bundles, which follows the "legacy" convention, so that's one case we can easily cover. I don't think we need to worry about the edge case, unless requested by someone, at the same time it's not that complicated if we wanted to support it. For the modern behavior, I think we could defer until someone asks for it.

Reply:
Should this bundle layout logic live in rules_swift?

Reply:
I filed FB7221978 for adding .swiftmodule directory support to ld.

Reply:
Even though the compiler supports .swiftmodule directories in general `-I` search paths, they're mostly used by Apple platforms for framework bundles and I don't believe any other platforms use fat libraries in the same way. So I don't think there's much gained by splitting a small part of this change across two repositories, when rules_swift otherwise wouldn't use that code.